pymc3/step_methods/smc_utils.py

```diff
@@ -78,3 +85,197 @@ def _posterior_to_trace(posterior, variables, model, var_info):
             size += new_size
         strace.record({k: v for k, v in zip(varnames, value)})
     return MultiTrace([strace])
+
+
+def metrop_kernel(
+    q_old,
+    old_tempered_logp,
+    proposal,
+    scaling,
+    accepted,
+    any_discrete,
+    all_discrete,
+    discrete,
+    n_steps,
+    prior_logp,
+    likelihood_logp,
+    beta,
+    ABC,
+):
+    """
+    Metropolis kernel
+    """
+    deltas = np.squeeze(proposal(n_steps) * scaling)
+    for n_step in range(n_steps):
+        delta = deltas[n_step]
+
+        if any_discrete:
+            if all_discrete:
+                delta = np.round(delta, 0).astype("int64")
+                q_old = q_old.astype("int64")
+                q_new = (q_old + delta).astype("int64")
+            else:
+                delta[discrete] = np.round(delta[discrete], 0)
+                q_new = floatX(q_old + delta)
+        else:
+            q_new = floatX(q_old + delta)
+
+        ll = likelihood_logp(q_new)
+
+        new_tempered_logp = prior_logp(q_new) + ll * beta
+
+        q_old, accept = metrop_select(new_tempered_logp - old_tempered_logp, q_new, q_old)
+        if accept:
+            accepted += 1
+            old_tempered_logp = new_tempered_logp
+
+    return q_old, accepted
+
+
+def calc_beta(beta, likelihoods, threshold=0.5):
+    """
+    Calculate next inverse temperature (beta) and importance weights based on current beta
+    and tempered likelihood.
+
+    Parameters
+    ----------
+    beta : float
+        tempering parameter of current stage
+    likelihoods : numpy array
+        likelihoods computed from the model
+    threshold : float
+        Determines the change of beta from stage to stage, i.e.indirectly the number of stages,
+        the higher the value of threshold the higher the number of stage. Defaults to 0.5.
+        It should be between 0 and 1.
+
+    Returns
+    -------
+    new_beta : float
+        tempering parameter of the next stage
+    old_beta : float
+        tempering parameter of the current stage
+    weights : numpy array
+        Importance weights (floats)
+    sj : float
+        Partial marginal likelihood
+    """
+    low_beta = old_beta = beta
+    up_beta = 2.0
+    rN = int(len(likelihoods) * threshold)
+
+    while up_beta - low_beta > 1e-6:
+        new_beta = (low_beta + up_beta) / 2.0
+        weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+        weights = weights_un / np.sum(weights_un)
+        ESS = int(1 / np.sum(weights ** 2))
+        if ESS == rN:
+            break
+        elif ESS < rN:
+            up_beta = new_beta
+        else:
+            low_beta = new_beta
+    if new_beta >= 1:
+        new_beta = 1
+    sj = np.exp((new_beta - old_beta) * likelihoods)
+    weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+    weights = weights_un / np.sum(weights_un)
+    return new_beta, old_beta, weights, np.mean(sj)
+
+
+def logp_forw(out_vars, vars, shared):
```

Comment:
Don't we have a similar function elsewhere that does this?

Reply:
I think this is the only one.